src/buildstream_plugins/sources/cargo.py

```diff
@@ -294,6 +310,268 @@ def _get_mirror_file(self, sha=None):
+    # Get whether the current git repo is resolved
+    #
+    # Returns:
+    #   (bool): Always true since we always have a commit
+    #
+    def is_resolved(self):
+        return True
+
+    ########################################################
+    #                   Private helpers                    #
+    ########################################################
+
+    # _get_mirror_dir()
+    #
+    # Gets the local mirror directory for this upstream git repository
+    #
+    def _get_mirror_dir(self):
+        if self.repo.endswith(".git"):
+            norm_url = self.repo[:-4]
+        else:
+            norm_url = self.repo
+
+        return os.path.join(
+            self.cargo.get_mirror_directory(),
+            utils.url_directory_name(norm_url) + ".git",
+        )
+
+    # _get_url()
+    #
+    # Gets the git URL to download this crate from
+    #
+    # Args:
+    #    alias (str|None): The URL alias to apply, if any
+    #
+    # Returns:
+    #    (str): The URL for this crate
+    #
+    def _get_url(self, alias=None):
+        return self.cargo.translate_url(self.cargo.repo, alias_override=alias)
+
+    # _mirror_repo()
+    #
+    # Returns the mirror repo, initialized if it doesn not exist yet
+    #
+    # Returns:
+    #    (Repo): The mirror repo crate
+    #
+    def _mirror_repo(self):
+        try:
+            return Repo.init_bare(self._get_mirror_dir(), mkdir=True)
+        except FileExistsError:
+            return Repo(self._get_mirror_dir(), bare=True)
+
+
+# config_inherit_workspace()
+#
+# Adds inherited values to config
+#
+# Args:
+#    config (dict): Crate config
+#    workspace_config (dict): Workspace config
+def config_inherit_workspace(config, workspace_config):
+    workspace_deps = workspace_config.get("dependencies")
+    if workspace_deps is not None:
+        dependencies = []
+        for key in ["dependencies", "dev-dependencies", "build-dependencies"]:
+            if key in config:
+                dependencies.append(config[key])
+
+        if "traget" in config:
+            for target in config["target"].values():
+                if "dependencies" in target:
+                    dependencies.append(target["dependencies"])
+
+        for deps in dependencies:
+            inherit_deps(deps, workspace_deps)
+
+    workspace_package = workspace_config.get("package")
+    if workspace_package is not None:
+        inherit_package(config["package"], workspace_package)
+
+
+# inherit_package()
+#
+# Adds inherited [package] entries to config
+#
+# Args:
+#    items (dict): Package items
+#    workspace_items (dict): Workspace items
+def inherit_package(items, workspace_items):
+    for key, value in items.items():
+        if isinstance(value, dict) and "workspace" in value:
+            workspace_value = workspace_items.get(key)
+            if workspace_value is None:
+                raise SourceError("Can't inherit package information from workspace: Value missing from workspace")
```

Comment:
As someone not familiar with the intricacies of Cargo, this error message makes no sense to me sorry.